ext/case_transform/src/lib.rs

```diff
@@ -2,80 +2,128 @@
-            let newKey = RString::new(block(key.unwrap().to_string()));
-            result.store(newKey, newValue);
-        });
+        RString::new(&result)).to_any_object()
+    }
+}
+
+impl Transform for Symbol {
+    fn transform(&self, transform_function: &Fn(String) -> String) -> AnyObject {
+        let result = transform_function(self.to_string());
 
-        result
+        Symbol::new(&result).to_any_object()
     }
+}
+
+impl Transform for Hash {
+    fn transform(&self, transform_function: &Fn(String) -> String) -> AnyObject {
+        let mut result = Hash::new();
+
+        self.each(|key, value| {
+            let new_key = transform(key, transform_function);
+            let new_value = match value.ty() {
+                ValueType::Hash => transform(value, transform_function),
+                _ => value,
+            };
 
-    fn transformArray(value: Array, transformMethod: &Fn(AnyObject) -> AnyObject) -> Array {
-        value.map(|item| transformMethod(item)).unwrap()
+            result.store(new_key, new_value);
+        });
+
+        result.to_any_object()
     }
+}
+
+impl Transform for Array {
+    fn transform(&self, transform_function: &Fn(String) -> String) -> AnyObject {
+        // Temp hack to consume &self for iterator
+        let result = unsafe { self.to_any_object().to::<Array>() };
 
-    fn transformHash(value: Hash, transformMethod: &Fn(AnyObject) -> AnyObject) -> Hash {
-        deepTransformKeys(value, |key| transformMethod(key))
+        result.into_iter()
+            .map(|item| transform(item, transform_function))
+            .collect::<Array>()
+            .to_any_object()
     }
+}
 
-    fn transformSymbol(value: Symbol, transformMethod: &Fn(AnyObject) -> AnyObject) -> Symbol {
-        let transformed = transformMethod(value);
-        Symbol::new(transformed);
+trait TryTransform: Object {
+    fn try_transform<T>(&self,
+                        transform_function: &Fn(String) -> String)
+                        -> Result<AnyObject, RuruError>
+        where T: VerifiedObject + Transform
+    {
+        self.try_convert_to::<T>().map(|object| object.transform(transform_function))
     }
+}
+
+impl TryTransform for AnyObject {}
 
-    fn transform(
-        value: AnyObject,
-        objectTransform: &Fn(AnyObject) -> AnyObject,
-        keyTransform: &Fn(String) -> String
-    ) -> AnyObject {
-        match value.unwrap().ty() {
-            ValueType::Array => transformArray(value, objectTransform).to_any_object(),
-            ValueType::Hash => transformHash(value, objectTransform).to_any_object(),
-            ValueType::Symbol => transformSymbol(value, objectTransform).to_any_object(),
-            ValueType::RString => keyTransform(value).to_any_object(),
-            ValueType::Object => value
-        }
+fn transform(object: AnyObject, key_transform: &Fn(String) -> String) -> AnyObject {
+    let result = object.try_transform::<RString>(key_transform)
```

Comment:
Try to transform the object as a `RString`
Otherwise try to transform the object as a `Symbol`
...
Otherwise try to transform the object as a `AnyObject` (it is any unkown class, do nothing)